addons/sale/controllers/portal.py

```diff
@@ -324,11 +325,13 @@ def portal_quote_decline(self, order_id, access_token=None, decline_message=None
             # read directly during the flush due to access rights, necessitating manual caching.
             order_sudo.order_line.currency_id
 
-            _message_post_helper(
-                'sale.order',
-                order_sudo.id,
-                decline_message,
-                token=access_token,
+            if not request.env["sale.order"]._check_thread_portal_access(order_sudo.id, token=access_token):
+                raise Forbidden()
```

Comment:
Why is this check necessary ? What does it check that `_document_check_access` doesn't check ?
Seems strange to me to potentially raise a Forbidden in this step of the code, considering we already checked the access to the SO before.

Reply:
You're right. The order is already sudo in the former lines here. There is no need to double check.

Reply:
Although `_check_thread_portal_access` finds and adds the right partner to context too. This part was done via the `_message_post_helper` before, and the double check for portal access was inside that function. I suggest we use the `_check_thread_portal_access` instead of `_document_check_access` so we have both security check and correct partner ID to post a message.

Reply:
It's probably possible to merge those functions in another PR if we decide to do that, however for now, the only part of the previous `_message_post_helper` that we need here, is finding the correct partner. I extracted this part to a separate function so it would be accessible without double checking the security.